File: src/participant_id.R
#' Standardize participant IDs to PREFIX_NN (two-digit numeric suffix, underscore).
#'
#' Examples: OPT-7, OPT_7, OPT7, and opt_07 become OPT_07. Values that do not
#' match PREFIX + optional -/_ + digits (e.g. CONTROL) are returned uppercased
#' and trimmed unchanged.
#'
#' @param x Character vector (or coercible to character).
#' @return Character vector of the same length as `x`.
normalize_participant_id <- function(x) {
  if (length(x) == 0L) {
    return(x)
  }
  x_chr <- toupper(trimws(as.character(x)))
  vapply(
    seq_along(x_chr),
    function(i) {
      xi <- x_chr[i]
      if (is.na(x[i]) || is.na(xi)) {
        return(NA_character_)
      }
      if (!nzchar(xi)) {
        return(xi)
      }
      if (!grepl("^[A-Z]+[-_]?[0-9]+$", xi)) {
        return(xi)
      }
      parts <- regmatches(xi, regexec("^([A-Z]+)[_-]?([0-9]+)$", xi))[[1]]
      if (length(parts) < 3L) {
        return(xi)
      }
      num <- parts[3]
      if (nchar(num) == 1L) {
        num <- sprintf("%02d", as.integer(num))
      }
      paste0(parts[2], "_", num)
    },
    FUN.VALUE = character(1L),
    USE.NAMES = FALSE
  )
}

File: src/participant_id.py
"""Shared participant ID normalization for Python cleaning scripts."""

from __future__ import annotations

import math
import re


def normalize_participant_id_val(val):
    """
    Standardize OPT-style IDs to OPT_NN (two-digit suffix, underscore).

    Accepts OPT-7, OPT_7, OPT7, OPT_07, OPT_02_T0 (after stripping _T in caller).
    Non-matching values (e.g. CONTROL) are returned uppercased and trimmed.
    """
    if val is None:
        return val
    try:
        import pandas as pd

        if pd.isna(val):
            return val
    except ImportError:
        if isinstance(val, float) and math.isnan(val):
            return val
    except (TypeError, ValueError):
        pass
    s = re.sub(r"\s+", "", str(val).upper().strip())
    if not s:
        return val
    m = re.fullmatch(r"([A-Z]+)[_-]?([0-9]+)", s)
    if not m:
        return s
    prefix, num = m.group(1), m.group(2)
    if len(num) == 1:
        num = f"0{num}"
    return f"{prefix}_{num}"
